File: programmers/카카오 다시풀기/level2/프렌즈4블록.cpp
/*
    처음에 문제를 잘못 이해해서 잘못 구현을 했다.
    시뮬레이션 문제였다.
*/
#include <string>
#include <vector>

using namespace std;

int solution(int m, int n, vector<string> board) {
    int answer = 0;
    bool flag = true;
    
    while(flag) {
        vector<vector<bool> > visited(m, vector<bool>(n, false)); // 방문 처리 
        flag = false;
        
        // 블럭 작동 
        for(int i = 0; i < m-1; i++) {
            for(int j = 0; j < n-1; j++) {
                char block = board[i][j];
                
                if(board[i][j] == ' ') continue;
                
                if(block == board[i][j+1] && 
                  block == board[i+1][j] && 
                  block == board[i+1][j+1]) {
                    visited[i][j] = true;
                    visited[i][j+1] = true;
                    visited[i+1][j] = true;
                    visited[i+1][j+1] = true;
                    flag = true;
                }
            }
        }
        
        // 삭제된 블럭 수 세기 
        // 삭제된 블럭 내리기
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(visited[i][j]) {
                    answer++;
                    
                    // 한 칸 내리기
                    for(int k = i-1; k >= 0; k--) {
                        board[k+1][j] = board[k][j]; 
                        board[k][j] = ' ';
                    }
                }
            }
        }
    }
    
    return answer;
}

File: kakao_java/blind2018/프렌즈4블록.java
package blind2018;

public class 프렌즈4블록 {
    private char[][] map;
    private boolean[][] remove;

    public int solution(int m, int n, String[] board) {
        map = new char[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                map[i][j] = board[i].charAt(j);
            }
        }

        remove = new boolean[m][n];

        int count = 0;
        while (true) {
            boolean isFinish = true;

            for (int i = 0; i < m - 1; i++) {
                for (int j = 0; j < n - 1; j++) {
                    char block = map[i][j];
                    if (block == '.') {
                        continue;
                    }

                    if (block == map[i][j + 1] && block == map[i + 1][j] && block == map[i + 1][j + 1]) {
                        remove[i][j] = true;
                        remove[i][j + 1] = true;
                        remove[i + 1][j] = true;
                        remove[i + 1][j + 1]= true;
                        isFinish = false;
                    }
                }
            }

            if (isFinish) {
                break;
            }

            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (remove[i][j]) {
                        map[i][j] = '.';
                        remove[i][j] = false;
                        count++;
                    }
                }
            }

            for (int i = 0; i < n; i++) {
                for (int j = m - 1; j > 0; j--) {
                    if (map[j][i] == '.') {
                        int k = j;
                        while (map[k][i] == '.') {
                            k--;
                            if (k < 0) {
                                k = 0;
                                break;
                            }
                        }

                        map[j][i] = map[k][i];
                        map[k][i] = '.';
                    }
                }
            }

        }


        return count;
    }
}
